RHDEVDOCS-3918 - Document Alertmanager for user workload monitoring

bburt-rh opened the pull request:
Summary: This PR documents how to enable and configure Alertmanager for user workload monitoring (user-defined projects).

- Aligned team: DevTools
- For branches: OCP. 4.11+
- Jira: https://issues.redhat.com/browse/RHDEVDOCS-3918
- Direct link to doc previews (must be connected to RH VPN): 
- - Main content: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/enabling-alert-routing-for-user-defined-projects.html
- - New procedure based on old procedure: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/managing-alerts.html#applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing_managing-alerts
- - Minor update: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/monitoring-overview.html#components-for-monitoring-user-defined-projects_monitoring-overview
- - Minor update: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/managing-alerts.html#creating-alert-routing-for-user-defined-projects_managing-alerts
- SME review: @simonpasquier (approved)
- QE review: @juzhao (approved)
- Peer review: @rolfedh (approved)

simonpasquier commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
+= Enabling an Alertmanager instance for user-defined projects
+
+A cluster administrator can enable a dedicated instance of Alertmanager to use with user-defined projects in a cluster.
+Doing so enables developers and other non-administrator users of user-defined projects to set up individual alerting configurations to suit their needs without the help of an administrator.
```
It's still possible for developers to configure alert notifications themselves without the dedicated Alertmanager. The difference is that the same Alertmanager will be used for both platform and user-defined alerts which may not be a desired situation for cluster admins.

bburt-rh replied:
I've rewritten the introduction to try and better capture the use case and benefits of the separate Alertmanager instance

simonpasquier commented on modules/monitoring-setting-up-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -1,13 +1,15 @@
-As a cluster administrator, you can enable alert routing for user-defined projects. After doing so, you can allow users to configure alert routing for their user-defined projects. Users can then create and configure user-defined alert routing by creating or editing the `AlertmanagerConfig` objects.
+As a cluster administrator, you can set up alert routing for user-defined projects. 
+After doing so, you can allow users to configure alert routing for their user-defined projects using an Alertmanager instance dedicated to user workload monitoring. 
+Users can then create and configure user-defined alert routing by creating or editing the `AlertmanagerConfig` objects for their user-defined projects without the help of an administrator.
```
I think that we need an update below too. There are 2 possible configurations:
1. No dedicated Alertmanager for user-defined alerts -> the admin should add `alertmanagerMain: {enableUserAlertmanagerConfig: true}` to the `cluster-monitoring-config` configmap in the `openshift-monitoring` namespace.
2. Dedicated Alertmanager for user-defined alerts -> the admin should add `alertmanager: {enabled: true, enableUserAlertmanagerConfig: true}` to the `user-workload-monitoring-config` configmap in the `openshift-user-workload-monitoring` namespace.

juzhao replied:
for the second configuration, from https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L578-L587
```
// AlertmanagerUserWorkloadConfig defines the Alertmanager configuration for UWM.
type AlertmanagerUserWorkloadConfig struct {
	Enabled                  bool                                 `json:"enabled"`
	EnableAlertmanagerConfig bool                                 `json:"enableAlertmanagerConfig"`
	LogLevel                 string                               `json:"logLevel"`
	NodeSelector             map[string]string                    `json:"nodeSelector"`
	Tolerations              []v1.Toleration                      `json:"tolerations"`
	Resources                *v1.ResourceRequirements             `json:"resources"`
	VolumeClaimTemplate      *monv1.EmbeddedPersistentVolumeClaim `json:"volumeClaimTemplate"`
}
```
`the admin should add alertmanager: {enabled: true, enableUserAlertmanagerConfig: true}`
should be
`the admin should add alertmanager: {enabled: true, enableAlertmanagerConfig: true}`
verified in https://bugzilla.redhat.com/show_bug.cgi?id=2099937#c1

bburt-rh replied:
I added an extra procedure for enabling user-defined alerts for the default platform Alertmanager instance.

simonpasquier commented on modules/monitoring-creating-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -21,6 +13,7 @@ If you are a non-administrator user who has been given the `alert-routing-edit`
 
 * A cluster administrator has enabled monitoring for user-defined projects.
 * A cluster administrator has enabled alert routing for user-defined projects.
+* A cluster administrator has enabled Alertmanager for user-defined projects.
```
As written elsewhere, there could be a dedicated Alertmanager for user-defined alerts in the `openshift-user-workload-monitoring` namespace or we reuse the instance existing in the `openshift-monitoring` namespace.

bburt-rh replied:
I've deleted this as a prerequisite.

simonpasquier reviewed (commented):
Thanks! This looks pretty good to me. My main remark is about making clear that the dedicated Alertmanager is an option.

juzhao commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have enabled monitoring for user-defined projects.
+* You have enabled alert routing for user-defined projects.
```
is it needed? @simonpasquier

bburt-rh replied:
Deleted.

juzhao commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
+  namespace: openshift-user-workload-monitoring
+data:
+  config.yaml: |
+    alertmanager: true <1>
```
should be
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    alertmanager:
      enabled: true
```

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
++
+[source,terminal]
+----
+#  oc -n openshift-user-workload-monitoring get alertmanager
```
remove additional space after `#`
should be
`# oc -n openshift-user-workload-monitoring get alertmanager`

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
++
+[source,terminal]
+----
+#  oc -n openshift-user-workload-monitoring get svc
```
remove additional space after `#`
should be
`# oc -n openshift-user-workload-monitoring get svc`

juzhao commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
+prometheus-user-workload-thanos-sidecar   ClusterIP   None             <none>        10902/TCP                     112m
+thanos-ruler                              ClusterIP   172.30.76.238    <none>        9091/TCP,10901/TCP            112m
+thanos-ruler-operated                     ClusterIP   None             <none>        10902/TCP,10901/TCP           111m
+----
```
I am kindly think the it only show how to enable a dedicated instance of Alertmanager, the verification steps does not show how to prove the dedicated Alertmanager works
1. if enabled the dedicated Alertmanager, the user projects alerts are sent to this alertmanager instance only
2. if disabled the dedicated Alertmanager, the user projects alerts are sent to platform alertmanager(under openshift-monitoring)

simonpasquier replied:
I think that the "oc -n openshift-user-workload-monitoring get alertmanager" step should be enough as it will prove that the user didn't mess up the edition of the configmap.
If the Alertmanager resource is created but the Alertmanager pods can't be started for some reason (no node capacity, failed scheduling, ...), there would be alerts firing.

bburt-rh replied:
I've deleted all of the verification steps except for `oc -n openshift-user-workload-monitoring get alertmanager`

bburt-rh commented on modules/monitoring-enabling-an-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,103 @@
++
+[source,terminal]
+----
+$ oc -n openshift-monitoring edit configmap user-workload-monitoring-config
```
fixed

bburt-rh commented:
@juzhao @simonpasquier - Please review this PR that I've revised per your feedback. If you could also review the rendered output of the entire end-to-end procedure at http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-alerts-for-user-defined-projects.html that would be very helpful as well. 

Thank you both for your great feedback!

juzhao commented:
http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-the-monitoring-stack.html#configurable-monitoring-components_configuring-the-monitoring-stack
since we support UWM Alertmanager now, user-workload-monitoring-config config map key for Alertmanager is alertmanager, we should add to doc
```
Component | cluster-monitoring-config config map key | user-workload-monitoring-config config map key
Alertmanager | alertmanagerMain | alertmanager
```

simonpasquier commented on modules/monitoring-setting-up-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,28 @@
+
+[role="_abstract"]
+As a cluster administrator, you can set up alert routing for user-defined projects. 
+After doing so, you can allow users to configure alert routing for their user-defined projects using either the default Alertmanager instance or the optional Alertmanager instance dedicated only to user workload monitoring, if enabled. 
```
(nit) do we want to use "user workload monitoring" here? IIUC the official docs stick with "user-defined monitoring".

simonpasquier commented on modules/monitoring-setting-up-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,28 @@
+* For user-defined alerting rules, user-defined routing is scoped to the namespace in which the resource is defined.
+For example, a routing configuration in namespace `ns1` only applies to `PrometheusRules` resources in the same namespace.
+
+* The Cluster Monitoring Operator (CMO) does not deploy a second Alertmanager service dedicated to user-defined alerts.
```
I'm not too sure about the meaning of this sentence.

bburt-rh replied:
Deleting this sentence. I'm not sure why it was there from the previous release.

simonpasquier commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,42 @@
+    alertmanagerMain: 
+      enableUserAlertmanagerConfig: true <1>
+----
+<1> Set the `enableUserAlertmanagerConfig` value to `true` to allow users to create user-defined alerts that use the main platform instance of Alertmanager.
```
```suggestion
<1> Set the `enableUserAlertmanagerConfig` value to `true` to allow users to create user-defined alert routing configurtion that use the main platform instance of Alertmanager.
```

bburt-rh replied:
Changed per your suggestion

simonpasquier commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,42 @@
+  name: cluster-monitoring-config
+  namespace: openshift-monitoring
+data:
+  config.yaml: |
+    alertmanagerMain: 
+      enableUserAlertmanagerConfig: true <1>
```
```suggestion
  config.yaml: |
    enableUserWorkload: true
    alertmanagerMain: 
      enableUserAlertmanagerConfig: true <1>
```

bburt-rh replied:
added `enableUserWorkload: true` to the sample YAML

simonpasquier commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,42 @@
+[id="monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects_{context}"]
+= Enabling the platform Alertmanager instance for user-defined projects
+
+You can enable user-defined alerts to use the main Alertmanager instance to fire alerts for user-defined projects. 
```
```suggestion
You can enable user-defined alerts to use the main Alertmanager instance to send notifications for user-defined projects. 
```

bburt-rh replied:
changed per your suggestion

simonpasquier commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,42 @@
+----
+<1> Set the `enableUserAlertmanagerConfig` value to `true` to allow users to create user-defined alerts that use the main platform instance of Alertmanager.
++
+. Save the file to apply the changes. The main Alertmanager instance restarts automatically and applies the new configuration.
```
I'm not sure that the Alertmanager pods are going to be restarted.

juzhao replied:
not restarted

juzhao commented:
http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-alerts-for-user-defined-projects.html
configmap setting for
`Enabling alert routing for user-defined projects`
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
    alertmanagerMain:
      enableUserAlertmanagerConfig: true
```
and
configmap setting for
`Enabling the platform Alertmanager instance for user-defined projects`
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    alertmanagerMain:
      enableUserAlertmanagerConfig: true 
```
are duplicated, we could merge them to one paragraph, `enableUserWorkload: true` setting here is no need

bburt-rh commented on monitoring/configuring-alerts-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,44 @@
+toc::[]
+
+[role="_abstract"]
+In {product-title} {product-version}, a cluster administrator can set up alerts and alert routing for user-defined projects, which involves three general configuration steps:
```
fixed. Changed entire topic to focus on enabling the feature rather than configuring it.

bburt-rh commented on monitoring/configuring-alerts-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,44 @@
+
+* Enable alert routing for user-defined projects.
+* Grant users permission to configure alert routing for user-defined projects.
+* Enable alerts for user-defined projects using either the default platform Alertmanager instance or a separate Alertmanager instance only for user-defined projects.
```
changed to `Enable alert notifications`

bburt-rh commented:
> http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-alerts-for-user-defined-projects.html `Enabling alert routing for user-defined projects` and `Enabling the platform Alertmanager instance for user-defined projects` are duplicated, they are the same thing.
> 
> configmap setting for `Enabling alert routing for user-defined projects`
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: cluster-monitoring-config
>   namespace: openshift-monitoring
> data:
>   config.yaml: |
>     enableUserWorkload: true
>     alertmanagerMain:
>       enableUserAlertmanagerConfig: true
> ```
> 
> and configmap setting for `Enabling the platform Alertmanager instance for user-defined projects`
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: cluster-monitoring-config
>   namespace: openshift-monitoring
> data:
>   config.yaml: |
>     alertmanagerMain:
>       enableUserAlertmanagerConfig: true 
> ```
> 
> and `enableUserWorkload: true` setting here is no need to add

@juzhao I've deleted the duplicated content

juzhao commented:
http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-alerts-for-user-defined-projects.html
it is 404, do we have change for it?

juzhao commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,68 @@
+  namespace: openshift-user-workload-monitoring
+data:
+  config.yaml: |
+    enableUserWorkload: true
```
`enableUserWorkload: true` should be removed, since it is enabled  in the cluster-monitoring-config config map under the openshift-monitoring namespace, the Prerequisites already enabled user projects monitoring
`You have enabled monitoring for user-defined projects in the cluster-monitoring-config config map in the openshift-monitoring namespace.`

bburt-rh replied:
removed

juzhao commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,68 @@
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have enabled monitoring for user-defined projects in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.
```
`in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace`
`in the ** config map in the ** namespace`, the expression seems not good

juzhao commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,42 @@
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have enabled monitoring for user-defined projects in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.
```
`You have enabled monitoring for user-defined projects in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.`
should be removed

bburt-rh replied:
Removed

juzhao commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,68 @@
+      enableAlertmanagerConfig: true <2>
+----
+<1> Set the `enabled` value to `true` to enable a dedicated instance of the Alertmanager for user-defined projects in a cluster. Set the value to `false` or omit the key entirely to disable the Alertmanager for user-defined projects.
+<2> Set the `enableAlertmanagerConfig` value to `true` to allow users to create user-defined alerts.
```
Set the `enableAlertmanagerConfig` value to `true` to allow users to create user-defined alert routing configurations that use the dedicated platform instance of Alertmanager.

bburt-rh commented:
> http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/configuring-alerts-for-user-defined-projects.html it is 404, do we have change for it? or it should be http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/enabling-alert-routing-for-user-defined-projects.html

@juzhao Fixed

rolfedh commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,67 @@
+[id="monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects_{context}"]
+= Enabling a separate Alertmanager instance for user-defined projects
+ 
+In some clusters, you might not want to use the default platform instance of Alertmanager for user-defined alerts. 
+In these cases, you can optionally enable a dedicated instance of Alertmanager to send alert notifications for user-defined projects in a cluster.
```
The juxtaposition of "user-defined alerts" in line 9 and "alert notifications for user-defined projects in a cluster" in line 10 is raises questions. Are the two equivalent or distinct?

bburt-rh replied:
Edited to read `In some clusters, you might not want to use the default platform instance of Alertmanager to send alert notifications for user-defined projects. 
In these cases, you can optionally enable a dedicated instance of Alertmanager to send alerts for these projects.`

rolfedh commented on modules/monitoring-understanding-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -7,16 +7,24 @@
-As a cluster administrator, you can enable alert routing for user-defined projects. After doing so, you can allow users to configure alert routing for their user-defined projects. Users can then create and configure user-defined alert routing by creating or editing the `AlertmanagerConfig` objects.
+As a cluster administrator, you can enable alert routing for user-defined projects. 
+With this feature, you can allow users with the **alert-routing-edit** role to configure alert notification routing and receivers for user-defined projects.
+These notifications will then be routed using either the default Alertmanager instance or the optional Alertmanager instance dedicated only to user-defined monitoring, if enabled.
```
```suggestion
These notifications are routed by the default Alertmanager instance or, if enabled, by an optional Alertmanager instance dedicated to user-defined monitoring.
```
- Avoids future tense.
- Avoids retrospectively figuring out what "if enabled" applies to.
- Doesn't avoid passive voice.

bburt-rh replied:
Edited to read `These notifications are routed by the default Alertmanager instance or, if enabled, an optional Alertmanager instance dedicated to user-defined monitoring.` Much better!

rolfedh commented on monitoring/enabling-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -6,30 +6,32 @@ include::_attributes/common-attributes.adoc[]
-
 [role="_abstract"]
-In {product-title} {product-version}, a cluster administrator can enable alert routing for user-defined projects.
+In {product-title} {product-version}, a cluster administrator can enable alert  routing for user-defined projects. This process consists of two general steps:
```
```suggestion
In {product-title} {product-version}, a cluster administrator can enable alert routing for user-defined projects. This process consists of two general steps:
```

bburt-rh replied:
fixed

rolfedh reviewed (approved):
Approved. Left some optional suggestions.
Please get SME and QE approval before requesting merge.

bburt-rh commented on modules/monitoring-understanding-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -7,16 +7,24 @@
-After a user has defined alert routing for a user-defined project, user-defined alerts are routed to the `alertmanager-main` pods in the `openshift-monitoring` namespace. 
+Users can then create and configure user-defined alert routing by creating or editing the `AlertmanagerConfig` objects for their user-defined projects without the help of an administrator.
+
+
```
fixed

juzhao commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,67 @@
+$ oc -n openshift-user-workload-monitoring edit configmap user-workload-monitoring-config
+----
++
+. Add `enabled: true` and `enableAlertmanagerConfig: true` under `alertmanager` under `data/config.yaml`:
```
Add `enabled: true`  and `enableAlertmanagerConfig: true` for the `alertmanager` component under `data/config.yaml`:

bburt-rh replied:
Changed per your suggestion.

juzhao commented:
LGTM, waiting for others to review

simonpasquier commented on modules/monitoring-components-for-monitoring-user-defined-projects.adoc:
```diff
@@ -23,6 +23,9 @@
 |The Thanos Ruler is a rule evaluation engine for Prometheus that is deployed as a separate process. In {product-title} {product-version}, Thanos Ruler provides rule and alerting evaluation for the monitoring of user-defined projects.
 
+|Alertmanager
+|The Alertmanager service handles alerts received from Prometheus. Alertmanager is also responsible for sending alerts to external notification systems. Deploying this service is optional.
```
```suggestion
|The Alertmanager service handles alerts received from Prometheus. Alertmanager is also responsible for sending user-defined alerts to external notification systems. Deploying this service is optional.
```

bburt-rh replied:
fixed

bburt-rh commented on modules/monitoring-understanding-alert-routing-for-user-defined-projects.adoc:
```diff
@@ -7,16 +7,23 @@
 
-* The Cluster Monitoring Operator (CMO) does not deploy a second Alertmanager service dedicated to user-defined alerts.
-Cluster administrators continue to define the main Alertmanager configuration by using a custom secret or the {product-title} web console. 
+* Cluster administrators continue to define the main Alertmanager configuration by using a custom secret or the {product-title} web console. 
```
Changed to 

`*  If the user-defined Alertmanager instance is enabled, cluster administrators can also define the main configuration of the user-defined Alertmanager instance.`

bburt-rh commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,41 @@
+[id="monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects_{context}"]
+= Enabling the platform Alertmanager instance for user-defined projects
+
+You can enable the platform Alertmanager instance to manage alert notifications for user-defined projects. 
```
Changed to your suggestion

bburt-rh commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,67 @@
+      enabled: true <1>
+      enableAlertmanagerConfig: true <2>
+----
+<1> Set the `enabled` value to `true` to enable a dedicated instance of the Alertmanager for user-defined projects in a cluster. Set the value to `false` or omit the key entirely to disable the Alertmanager for user-defined projects.
```
Changed to:

`If you set this value to `false` or if the key is omitted, user-defined alerts are routed to the default platform Alertmanager instance`

simonpasquier commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,63 @@
+      enableAlertmanagerConfig: true <2>
+----
+<1> Set the `enabled` value to `true` to enable a dedicated instance of the Alertmanager for user-defined projects in a cluster. Set the value to `false` or omit the key entirely to disable the Alertmanager for user-defined projects. 
+If you set this value to `false` or if the key is omitted, user-defined alerts are routed to the default platform Alertmanager instance
```
missing period?

bburt-rh replied:
fixed

simonpasquier commented on modules/monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,41 @@
+
+:_content-type: PROCEDURE
+[id="monitoring-enabling-the-platform-alertmanager-instance-for-user-defined-projects_{context}"]
+= Enabling the platform Alertmanager instance for user-defined projects
```
should it be "Enabling the platform Alertmanager instance for user-defined alert routing"?

bburt-rh replied:
Revised per your suggestion

simonpasquier commented on modules/monitoring-enabling-a-separate-alertmanager-instance-for-user-defined-projects.adoc:
```diff
@@ -0,0 +1,63 @@
+----
+<1> Set the `enabled` value to `true` to enable a dedicated instance of the Alertmanager for user-defined projects in a cluster. Set the value to `false` or omit the key entirely to disable the Alertmanager for user-defined projects. 
+If you set this value to `false` or if the key is omitted, user-defined alerts are routed to the default platform Alertmanager instance
+<2> Set the `enableAlertmanagerConfig` value to `true` to enable users to define their own alert routing configurations that use the dedicated instance of Alertmanager rather than the platform instance.
```
```suggestion
<2> Set the `enableAlertmanagerConfig` value to `true` to enable users to define their own alert routing configurations with `AlertmanagerConfig` objects.
```

bburt-rh replied:
Changed per your suggestion

simonpasquier reviewed (commented):
We need to update the note [1] about the Alertmanager configuration secret resource to also mention the `openshift-user-workload-monitoring` namespace when user-defined Alertmanager is enabled.
There's also a section [2] in the current documentation about editing the openshift-monitoring's Alertmanager secret. Not sure if we need something identical for openshift-user-workload-monitoring? FWIW the procedure is exactly the same except the name (`alertmanager-user-workload` vs. `alertmanager-main`) and namespace.

[1] https://docs.openshift.com/container-platform/4.10/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack
[2] https://docs.openshift.com/container-platform/4.10/monitoring/managing-alerts.html#applying-custom-alertmanager-configuration_managing-alerts

bburt-rh commented:
> We need to update the note [1] about the Alertmanager configuration secret resource to also mention the `openshift-user-workload-monitoring` namespace when user-defined Alertmanager is enabled. There's also a section [2] in the current documentation about editing the openshift-monitoring's Alertmanager secret. Not sure if we need something identical for openshift-user-workload-monitoring? FWIW the procedure is exactly the same except the name (`alertmanager-user-workload` vs. `alertmanager-main`) and namespace.


@simonpasquier For your first point, I've edited the Note in the "Support considerations" section as follows:

```
The Alertmanager configuration is deployed as a secret resource in the `openshift-monitoring` namespace. 
If you have enabled a separate Alertmanager instance for user-defined alert routing, an Alertmanager configuration is also deployed as a secret resource in the `openshift-user-workload-monitoring` namespace. 
To configure additional routes for any instance of Alertmanager, you need to decode, modify, and then encode that secret. 
This procedure is a supported exception to the preceding statement.
```


> We need to update the note [1] about the Alertmanager configuration secret resource to also mention the `openshift-user-workload-monitoring` namespace when user-defined Alertmanager is enabled. There's also a section [2] in the current documentation about editing the openshift-monitoring's Alertmanager secret. Not sure if we need something identical for openshift-user-workload-monitoring? FWIW the procedure is exactly the same except the name (`alertmanager-user-workload` vs. `alertmanager-main`) and namespace.
> 
> [1] https://docs.openshift.com/container-platform/4.10/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack [2] https://docs.openshift.com/container-platform/4.10/monitoring/managing-alerts.html#applying-custom-alertmanager-configuration_managing-alerts

@simonpasquier 
1. I updated the note in the "Support considerations" section per your instructions. 
2. I added a new procedure for editing the secret for the dedicated UWM Alertmanager instance. It's pretty much identical to the procedure for platform monitoring except for the things you mentioned: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/managing-alerts.html#applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing_managing-alerts

simonpasquier commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,93 @@
++
+[source,yaml]
+----
+global:
+  resolve_timeout: 5m
+route:
+  group_wait: 30s
+  group_interval: 5m
+  repeat_interval: 12h
+  receiver: default
+  routes:
+  - match:
+      alertname: Watchdog
+    repeat_interval: 5m
+    receiver: watchdog
+  - match:
+      service: <your_service> <1>
+    routes:
+    - match:
+        <your_matching_rules> <2>
+      receiver: <receiver> <3>
+receivers:
+- name: default
+- name: watchdog
+- name: <receiver>
+#  <receiver_configuration>
```
In practice, the default configuration is much simpler than that. We can also avoid the nested `routes` field to keep it concise.

```suggestion
route:
  receiver: Default
  group_by:
  - name: Default
  routes:
  - matchers: <1>
    - "service = prometheus-example-monitor"
    receiver: <receiver> <2>
receivers:
- name: Default
- name: <receiver>
#  <receiver_configuration>
```

bburt-rh replied:
Changed per your suggestion

simonpasquier commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,93 @@
+#  <receiver_configuration>
+----
+<1> `service` specifies the service that fires the alerts.
+<2> `<your_matching_rules>` specifies the target alerts.
```
this would be dropped if applying my suggestion above

simonpasquier commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,93 @@
+- name: <receiver>
+#  <receiver_configuration>
+----
+<1> `service` specifies the service that fires the alerts.
```
```suggestion
<1> specifies which alerts match the route, in this example all alerts which have the `service="prometheus-example-monitor"` label.
```

bburt-rh replied:
Edited per your suggestion

simonpasquier commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,93 @@
+----
+<1> `service` specifies the service that fires the alerts.
+<2> `<your_matching_rules>` specifies the target alerts.
+<3> `receiver` specifies the receiver to use for the alert.
```
```suggestion
<3> `receiver` specifies the receiver to use for the alerts group.
```

bburt-rh replied:
Revised per your addition

simonpasquier commented on modules/monitoring-components-for-monitoring-user-defined-projects.adoc:
```diff
@@ -23,6 +23,9 @@
 |The Thanos Ruler is a rule evaluation engine for Prometheus that is deployed as a separate process. In {product-title} {product-version}, Thanos Ruler provides rule and alerting evaluation for the monitoring of user-defined projects.
 
+|Alertmanager
+|The Alertmanager service handles alerts received from Prometheus. Alertmanager is also responsible for sending user-defined alerts to external notification systems. Deploying this service is optional.
```
```suggestion
|The Alertmanager service handles alerts received from Prometheus and Thanos Ruler. Alertmanager is also responsible for sending user-defined alerts to external notification systems. Deploying this service is optional.
```

bburt-rh replied:
Added

simonpasquier commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,93 @@
++
+The following Alertmanager configuration example configures PagerDuty as an alert receiver:
++
+[source,yaml,subs=quotes]
+----
+global:
+  resolve_timeout: 5m
+route:
+  group_wait: 30s
+  group_interval: 5m
+  repeat_interval: 12h
+  receiver: default
+  routes:
+  - match:
+      alertname: Watchdog
+    repeat_interval: 5m
+    receiver: watchdog
+  *- match:
+      service: example-app
+    routes:
+    - match:
+        severity: critical
+      receiver: team-frontend-page*
+receivers:
+- name: default
+- name: watchdog
+*- name: team-frontend-page
+  pagerduty_configs:
+  - service_key: "_your-key_"*
+----
++
+With this configuration, alerts of `critical` severity fired by the `example-app` service are sent using the `team-frontend-page` receiver. Typically, these types of alerts are paged to an individual or a critical response team.
```
I think that we can keep only one concrete example of modified Alertmanager configuration and remove this part.

bburt-rh replied:
removed

simonpasquier reviewed (commented):
lgtm

bburt-rh commented:
@juzhao Please review this topic: http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/managing-alerts.html#applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing_managing-alerts. If no changes are needed, please give an approval as well. Thank you!

juzhao commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,49 @@
+
+.Procedure
+
+. Print the currently active Alertmanager configuration into the file `alertmanager.yaml`:
```
since we already have file named `alertmanager.yaml`, 
http://file.rdu.redhat.com/bburt/RHDEVDOCS-3918-alertmanager-config-for-user-defined-monitoring/monitoring/managing-alerts.html#applying-custom-alertmanager-configuration_managing-alerts
it's better to change to `user-workload-alertmanager.yaml` to distinguish

juzhao commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,49 @@
++
+[source,terminal]
+----
+$ oc -n openshift-user-workload-monitoring get secret alertmanager-user-workload --template='{{ index .data "alertmanager.yaml" }}' | base64 --decode > alertmanager.yaml
```
`$ oc -n openshift-user-workload-monitoring get secret alertmanager-user-workload --template='{{ index .data "alertmanager.yaml" }}' | base64 --decode > user-workload-alertmanager.yaml`

juzhao commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,49 @@
+$ oc -n openshift-user-workload-monitoring get secret alertmanager-user-workload --template='{{ index .data "alertmanager.yaml" }}' | base64 --decode > alertmanager.yaml
+----
++
+. Edit the configuration in `alertmanager.yaml`:
```
user-workload-alertmanager.yaml

juzhao commented on modules/monitoring-applying-a-custom-configuration-to-alertmanager-for-user-defined-alert-routing.adoc:
```diff
@@ -0,0 +1,49 @@
++
+[source,terminal]
+----
+$ oc -n openshift-user-workload-monitoring create secret generic alertmanager-user-workload --from-file=alertmanager.yaml --dry-run=client -o=yaml |  oc -n openshift-user-workload-monitoring replace secret --filename=-
```
`$ oc -n openshift-user-workload-monitoring create secret generic alertmanager-user-workload --from-file=user-workload-alertmanager.yaml --dry-run=client -o=yaml |  oc -n openshift-user-workload-monitoring replace secret --filename=-`

juzhao commented:
/lgtm

abrennan89 commented:
/cherrypick enterprise-4.11
